File: examples/cxx/test_functions/maximize_function.cpp
#include "korali.h"
#include "models/ackley.h"
#include "models/eggholder.h"
#include "models/levi13.h"
#include "models/rosenbrock.h"

int main(int argc, char* argv[])
{

  if(argc<2){
    printf("\nInvalid Args! Usage:\n\t./maximize_function 1 (=maximize -ackley)\n\t./maximize_function 2 (=maximize -eggholder)\n\t./maximize_function 3 (=maximize -levi13)\n\t./maximize_function 4 (=maximize -rosenbrock)\n\n");
    exit(1);
  }

 std::function<void(Korali::modelData&)> model;

 int nParams;
 double lower, upper;
 switch( atoi(argv[1]) ){
    case 1:
        nParams = 4; lower = -32.0; upper = 32.0;
        model = [](Korali::modelData& d) { m_ackley(d.getParameters(), d.getResults()); }; 
        break;

    case 2:
        nParams = 2; lower = -512.0; upper = 512.0;
        model = [](Korali::modelData& d) { m_eggholder(d.getParameters(), d.getResults()); }; 
        break;

    case 3:
        nParams = 2; lower = -10.0; upper = 10.0;
        model = [](Korali::modelData& d) { m_levi13(d.getParameters(), d.getResults()); }; 
        break;

    case 4: 
        nParams = 2; lower = -32.0; upper = 32.0;
        model = [](Korali::modelData& d) { m_rosenbrock(d.getParameters(), d.getResults()); }; break;

 }

 auto korali = Korali::Engine(model);

 korali["Seed"] = 0xC0FFEE;
 korali["Verbosity"] = "Detailed";

 for (int i = 0; i < nParams; i++)
 {
  korali["Parameters"][i]["Name"] = "X" + std::to_string(i);
  korali["Parameters"][i]["Type"] = "Computational";
  korali["Parameters"][i]["Distribution"] = "Uniform";
  korali["Parameters"][i]["Minimum"] = lower;
  korali["Parameters"][i]["Maximum"] = upper;
 }

 korali["Problem"]["Objective"] = "Direct Evaluation";

 korali["Solver"]["Method"] = "CMA-ES";
 korali["Solver"]["Lambda"] = 8;
 korali["Solver"]["Termination Criteria"]["Min DeltaX"] = 1e-12;
 
 korali.run();

 return 0;
}

File: examples/python/test_functions/maximize_function.py
#!/usr/bin/env python3
import sys
import threading
sys.path.append('./model')
from ackley import m_ackley
from gaussian import gaussian
from eggholder import m_eggholder
from levi13 import m_levi13
from rosenbrock import m_rosenbrock
from shekel import m_shekel

import libkorali

error = """\nInvalid Args! Usage:\n
            \t./maximize_function 1 (=maximize -ackley)\n
            \t./mazimize_function 2 (=mazimize -eggholder)\n
            \t./mazimize_function 3 (=mazimize +gaussian)\n                   
            \t./mazimize_function 4 (=mazimize -levi13)\n                       
            \t./mazimize_function 5 (=mazimzie -rosenbrock)\n                   
            \t./mazimize_function 6 (=maximize -shekel)\n\n"""

if len(sys.argv) != 2:
    print(error)
    exit(1)

case = int(sys.argv[1])

if (case == 1) :
    nParams = 4
    lower, upper = -32.0, 32.0
    korali = libkorali.Engine(m_ackley)

elif (case == 2) :
    nParams = 2
    lower, upper = -512.0, 512.0
    korali = libkorali.Engine(m_eggholder)

elif (case == 3) :
    nParams = 5
    lower, upper = -32.0, 32.0
    korali = libkorali.Engine(gaussian)

elif (case == 4) :
    nParams = 2
    lower, upper = -10.0, 10.0
    korali = libkorali.Engine(m_levi13)

elif (case == 5) :
    nParams = 2
    lower, upper = -32.0, 32.0
    korali = libkorali.Engine(m_rosenbrock)

elif (case == 6) :
    nParams = 4
    lower, upper = 0.0, 10.0
    korali = libkorali.Engine(m_shekel)

else :
    print(error)
    exit(1)

korali["Seed"]      = 0xC0FFEE
korali["Verbosity"] = "Detailed"

for i in range(nParams):
    korali["Problem"]["Variables"][i]["Name"] = "X" + str(i)
    korali["Problem"]["Variables"][i]["Type"] = "Computational"
    korali["Problem"]["Variables"][i]["Distribution"] = "Uniform"
    korali["Problem"]["Variables"][i]["Minimum"] = lower
    korali["Problem"]["Variables"][i]["Maximum"] = upper
     
korali["Problem"]["Type"] = "Direct"

korali["Solver"]["Method"] = "CMA-ES"
korali["Solver"]["Lambda"] = 10
korali["Solver"]["Termination Criteria"]["Max Generations"] = 100
korali["Solver"]["Termination Criteria"]["Min DeltaX"] = 1e-12

korali.run()
